### PR TITLE
fix `julia_type<void>`

### PR DESCRIPTION
**Pull request by mofeing**
fixes JuliaInterop/CxxWrap.jl#463

the erroring tests seem to be unrelated to the changed code

### CODE REVIEW

**Comment by barche** [comment]
Type `void` is already mapped here:
https://github.com/JuliaInterop/libcxxwrap-julia/blob/dd11c219fa837e1cb855e417e4eb95f1910c01b7/src/jlcxx.cpp#L397

So the first thing to find out is why this fails in your setup.

With regards to this PR, the error on e.g. Ubuntu is:

```
/home/runner/work/libcxxwrap-julia/libcxxwrap-julia/include/jlcxx/type_conversion.hpp:502:8: error: extra qualification not allowed [-fpermissive]
  502 | struct jlcxx::julia_type_factory<void>
      |        ^~~~~
```

So there is a syntax error. But adding a separate `julia_type_factory` for `void` is not the correct way to solve this, it just fights the symptom of a deeper problem.

**Comment by mofeing** [comment]
> With regards to this PR, the error on e.g. Ubuntu is:
> 
> ```
> /home/runner/work/libcxxwrap-julia/libcxxwrap-julia/include/jlcxx/type_conversion.hpp:502:8: error: extra qualification not allowed [-fpermissive]
>   502 | struct jlcxx::julia_type_factory<void>
>       |        ^~~~~
> ```
> 
> So there is a syntax error. But adding a separate `julia_type_factory` for `void` is not the correct way to solve this, it just fights the symptom of a deeper problem.

ah right, that is a typo sorry

> Type `void` is already mapped here:
> 
> https://github.com/JuliaInterop/libcxxwrap-julia/blob/dd11c219fa837e1cb855e417e4eb95f1910c01b7/src/jlcxx.cpp#L397
> 
> So the first thing to find out is why this fails in your setup.

yeah, that's weird. i even tried calling `register_core_types` and `register_core_cxxwrap_type` by myself inside the `define_module_julia` but doesn't work. 🫠

**Comment by mofeing** [comment]
okay, so i run this inside the function definining the module
```c++
std::cout << "has_julia_type<void>() = " << jlcxx::has_julia_type<void>() << std::endl;

jlcxx::register_core_types();
jlcxx::register_core_cxxwrap_types();

std::cout << "has_julia_type<void>() = " << jlcxx::has_julia_type<void>() << std::endl;
```

and it returns this during Julia precompilation
```
has_julia_type<void>() = 0
has_julia_type<void>() = 0
```

**Comment by barche** [comment]
I am not able to reproduce this, what kind of system are you running all this on, exactly?

**Comment by mofeing** [comment]
i managed to reproduce the error with a MWE and packaged everything here https://github.com/mofeing/ReproducerCxxWrapVoidBug

if you activate the package's env and call `]build` you will get the error when the package is precompiling. if you then comment this line, https://github.com/mofeing/ReproducerCxxWrapVoidBug/blob/299185e0aa5b9d6655c396dfbf89617b8b9e4329/deps/CMakeLists.txt#L18, the one adding the `-fvisibility=hidden`, the error will disappear (i believe `-fvisibility` is just for clang so if you use gcc, you need to use the equivalent).

AFAIK only the julia module defining function is `ccall`ed, but it seems like the addresses of the C++ function stored there are not correct or directly skipped. not sure about how this last part happens.

tldr: the problem is that we are filtering all symbols by default except some in a passlist, and the C++ symbols are not exported.

**Comment by mofeing** [comment]
so my problem was symbol visibility. due to some symbol conflict we had on macOS between a build of LLVM included in our lib and Julia's lib, we were hiding symbols by default and only exporting some symbols from our own C-API. i also exported the C-API of libcxxwrap_julia, but it seemed to not be enough.

the solution was to export all symbols by default and just hide the llvm symbols. probably you also need to export the C++ symbols of libcxxwrap. why it doesn't crash in a more obvious (like "hey, this symbol is missing") is out of my knowledge but symbol conflicts and visibility are hard to debug.
